File: scripts/fix-mkdocs-structure.py
#!/usr/bin/env python3
"""
Fix MkDocs structure by moving overview pages to index.md inside their directories.
This prevents duplicate navigation entries when a file and directory have the same name.
"""

import os
import shutil
from pathlib import Path

def fix_duplicate_entries(docs_dir):
    """
    Find markdown files that have a corresponding directory with the same name,
    and move them to index.md inside that directory.
    """
    fixes_made = []
    
    # Walk through all markdown files
    for root, dirs, files in os.walk(docs_dir):
        for file in files:
            if file.endswith('.md'):
                file_path = Path(root) / file
                basename = file[:-3]  # Remove .md extension
                dir_path = Path(root) / basename
                
                # Check if there's a directory with the same name
                if dir_path.exists() and dir_path.is_dir():
                    # Check if index.md already exists in that directory
                    index_path = dir_path / 'index.md'
                    if not index_path.exists():
                        # Move the file to index.md in the directory
                        shutil.move(str(file_path), str(index_path))
                        fixes_made.append(f"Moved {file_path.relative_to(docs_dir)} -> {index_path.relative_to(docs_dir)}")
                        
                        # Also move any associated images
                        for img_file in Path(root).glob(f"{basename}_*"):
                            if img_file.is_file():
                                img_dest = dir_path / img_file.name
                                shutil.move(str(img_file), str(img_dest))
                                fixes_made.append(f"Moved {img_file.relative_to(docs_dir)} -> {img_dest.relative_to(docs_dir)}")
                        
                        # Also move any images that match the pattern exactly
                        for img_ext in ['.png', '.jpg', '.jpeg', '.gif', '.svg']:
                            img_file = Path(root) / f"{basename}{img_ext}"
                            if img_file.exists():
                                img_dest = dir_path / img_file.name
                                shutil.move(str(img_file), str(img_dest))
                                fixes_made.append(f"Moved {img_file.relative_to(docs_dir)} -> {img_dest.relative_to(docs_dir)}")
    
    return fixes_made

def update_references(docs_dir):
    """
    Update references in markdown files to point to the new locations.
    """
    updates_made = []
    
    for root, dirs, files in os.walk(docs_dir):
        for file in files:
            if file.endswith('.md'):
                file_path = Path(root) / file
                content_changed = False
                
                with open(file_path, 'r', encoding='utf-8') as f:
                    content = f.read()
                    original_content = content
                
                # Update references to moved files
                # Look for markdown links like [text](path.md)
                import re
                from urllib.parse import unquote, quote
                pattern = r'\[([^\]]*)\]\(([^)]+\.md)\)'
                
                def fix_link(match):
                    text = match.group(1)
                    link = match.group(2)
                    
                    # Skip external links
                    if link.startswith('http'):
                        return match.group(0)
                    
                    # Decode URL-encoded paths for processing
                    decoded_link = unquote(link)
                    
                    # Resolve the link path relative to current file
                    current_dir = Path(root)
                    
                    # For any .md link, check if there's a directory with index.md
                    # that should be used instead
                    if not decoded_link.startswith('/'):
                        # Resolve relative to current directory
                        resolved_path = (current_dir / decoded_link).resolve()
                        
                        # Check if this points to a file that should be a directory
                        # Remove .md extension to get the potential directory name
                        if resolved_path.suffix == '.md':
                            potential_dir = resolved_path.with_suffix('')
                            potential_index = potential_dir / 'index.md'
                            
                            # If a directory with index.md exists, update the link
                            if potential_index.exists():
                                # Calculate relative path from current file to the directory
                                new_path = os.path.relpath(potential_dir, current_dir)
                                # Return link pointing to directory (MkDocs will serve index.md)
                                # Replace backslashes with forward slashes for consistency
                                new_path = new_path.replace('\\', '/')
                                # Re-encode spaces in the path for URL compatibility
                                new_path = new_path.replace(' ', '%20')
                                return f'[{text}]({new_path}/)'
                    
                    # Also handle local references (same directory)
                    elif '/' not in decoded_link:
                        basename = decoded_link[:-3]  # Remove .md
                        possible_dir = current_dir / basename
                        if possible_dir.exists() and possible_dir.is_dir():
                            # Re-encode spaces for URL compatibility
                            encoded_basename = basename.replace(' ', '%20')
                            return f'[{text}]({encoded_basename}/)'
                    
                    return match.group(0)
                
                content = re.sub(pattern, fix_link, content)
                
                if content != original_content:
                    with open(file_path, 'w', encoding='utf-8') as f:
                        f.write(content)
                    updates_made.append(f"Updated references in {file_path.relative_to(docs_dir)}")
    
    return updates_made

def main():
    # Get the docs directory
    script_dir = Path(__file__).parent
    project_root = script_dir.parent
    docs_dir = project_root / 'docs'
    
    if not docs_dir.exists():
        print(f"Error: docs directory not found at {docs_dir}")
        return 1
    
    print(f"Fixing MkDocs structure in {docs_dir}")
    print("-" * 50)
    
    # Fix duplicate entries
    fixes = fix_duplicate_entries(docs_dir)
    if fixes:
        print("Files reorganized:")
        for fix in fixes:
            print(f"  - {fix}")
    else:
        print("No duplicate entries found that need fixing")
    
    print()
    
    # Update references
    updates = update_references(docs_dir)
    if updates:
        print("References updated:")
        for update in updates:
            print(f"  - {update}")
    else:
        print("No references needed updating")
    
    print("-" * 50)
    print(f"Structure fix complete: {len(fixes)} files moved, {len(updates)} files updated")
    
    return 0

if __name__ == '__main__':
    exit(main())

File: scripts/fix-mkdocs-structure.ts
#!/usr/bin/env node
/**
 * Fix MkDocs structure by:
 * 1. Syncing README.md to docs/index.md with necessary path adjustments
 * 2. Moving overview pages to index.md inside their directories to prevent duplicate navigation entries
 */

import * as fs from 'fs';
import * as path from 'path';

interface FixResult {
    message: string;
}

/**
 * Find markdown files that have a corresponding directory with the same name,
 * and move them to index.md inside that directory.
 */
function fixDuplicateEntries(docsDir: string): FixResult[] {
    const fixesMade: FixResult[] = [];
    
    function walkDir(dir: string): void {
        let files: string[];
        try {
            files = fs.readdirSync(dir);
        } catch (err) {
            console.warn(`Warning: Unable to read directory ${dir}: ${err.message}`);
            return;
        }
        
        for (const file of files) {
            const filePath = path.join(dir, file);
            let stat: fs.Stats;
            
            try {
                stat = fs.statSync(filePath);
            } catch (err) {
                // File might have been moved already, skip it
                continue;
            }
            
            if (stat.isDirectory()) {
                walkDir(filePath);
            } else if (file.endsWith('.md')) {
                const basename = file.slice(0, -3); // Remove .md extension
                const dirPath = path.join(dir, basename);
                
                // Check if there's a directory with the same name
                if (fs.existsSync(dirPath) && fs.statSync(dirPath).isDirectory()) {
                    const indexPath = path.join(dirPath, 'index.md');
                    
                    // Check if index.md already exists in that directory
                    if (!fs.existsSync(indexPath)) {
                        // Move the file to index.md in the directory
                        fs.renameSync(filePath, indexPath);
                        fixesMade.push({
                            message: `Moved ${path.relative(docsDir, filePath)} -> ${path.relative(docsDir, indexPath)}`
                        });
                        
                        // Move associated images with pattern basename_*
                        try {
                            const dirFiles = fs.readdirSync(dir);
                            for (const imgFile of dirFiles) {
                                if (imgFile.startsWith(`${basename}_`)) {
                                    const imgSrc = path.join(dir, imgFile);
                                    try {
                                        if (!fs.statSync(imgSrc).isDirectory()) {
                                            const imgDest = path.join(dirPath, imgFile);
                                            fs.renameSync(imgSrc, imgDest);
                                            fixesMade.push({
                                                message: `Moved ${path.relative(docsDir, imgSrc)} -> ${path.relative(docsDir, imgDest)}`
                                            });
                                        }
                                    } catch (err) {
                                        // File might have been moved already, skip it
                                    }
                                }
                            }
                        } catch (err) {
                            // Directory might not exist anymore, skip it
                        }
                        
                        // Move exact match images
                        const imgExtensions = ['.png', '.jpg', '.jpeg', '.gif', '.svg'];
                        for (const ext of imgExtensions) {
                            const imgFile = path.join(dir, `${basename}${ext}`);
                            if (fs.existsSync(imgFile)) {
                                const imgDest = path.join(dirPath, `${basename}${ext}`);
                                fs.renameSync(imgFile, imgDest);
                                fixesMade.push({
                                    message: `Moved ${path.relative(docsDir, imgFile)} -> ${path.relative(docsDir, imgDest)}`
                                });
                            }
                        }
                    }
                }
            }
        }
    }
    
    walkDir(docsDir);
    return fixesMade;
}

/**
 * Update references in markdown files to point to the new locations.
 */
function updateReferences(docsDir: string): FixResult[] {
    const updatesMade: FixResult[] = [];
    
    function fixLink(match: string, text: string, link: string, currentDir: string, isIndex: boolean): string {
        // Skip external links, mailto, and special protocols
        if (link.startsWith('http') || link.startsWith('mailto:') || link.startsWith('xmpp:')) {
            return match;
        }
        
        // Skip anchor-only links
        if (link.startsWith('#')) {
            return match;
        }
        
        // Handle malformed links with nested brackets (e.g., [developers]([url](https://...))
        if (link.includes('[') || link.includes(']')) {
            // This is a malformed link, skip it
            return match;
        }
        
        // Handle links wrapped in angle brackets (e.g., <https://...>)
        if (link.startsWith('<') && link.endsWith('>')) {
            // This is likely a literal URL that shouldn't be processed
            return match;
        }
        
        // Decode URL-encoded paths for processing
        let decodedLink: string;
        try {
            decodedLink = decodeURIComponent(link);
        } catch (err) {
            // If decoding fails, use the original link
            decodedLink = link;
        }
        
        // Extract anchor if present
        let anchorPart = '';
        if (decodedLink.includes('#')) {
            const parts = decodedLink.split('#');
            decodedLink = parts[0];
            anchorPart = '#' + parts.slice(1).join('#');
        }
        
        // Special case: if we're in index.md and the link starts with the parent directory name
        // This happens when a file was converted to index.md and had links to siblings
        if (isIndex && decodedLink.includes('/')) {
            const pathParts = decodedLink.split('/');
            const parentDirName = path.basename(currentDir);
            
            // Check if first part matches the parent directory name
            if (pathParts[0] === parentDirName) {
                // This is a self-referential path, strip the first part
                const fixedLink = pathParts.slice(1).join('/');
                // Re-encode spaces for URL compatibility before recursing
                const fixedLinkEncoded = fixedLink.replace(/ /g, '%20');
                // Recursively process the fixed link
                return fixLink(`[${text}](${fixedLinkEncoded}${anchorPart})`, text, fixedLinkEncoded + anchorPart, currentDir, isIndex);
            }
        }
        
        // For any .md link, check if there's a directory with index.md
        // that should be used instead
        if (!decodedLink.startsWith('/')) {
            // Resolve relative to current directory
            const resolvedPath = path.resolve(currentDir, decodedLink);
            
            // Check if this points to a file that should be a directory
            // Remove .md extension to get the potential directory name
            if (resolvedPath.endsWith('.md')) {
                const potentialDir = resolvedPath.slice(0, -3);
                const potentialIndex = path.join(potentialDir, 'index.md');
                
                // If a directory with index.md exists, update the link
                if (fs.existsSync(potentialIndex)) {
                    // If we're in an index.md file and linking to a file that's now
                    // in a sibling directory, adjust the path
                    if (isIndex) {
                        // Check if they share the same parent directory
                        if (path.dirname(potentialDir) === path.dirname(currentDir)) {
                            // It's a sibling - just use directory name
                            const dirName = path.basename(potentialDir).replace(/ /g, '%20');
                            return `[${text}](${dirName}/${anchorPart})`;
                        }
                    }
                    
                    // Calculate relative path from current file to the directory
                    const newPath = path.relative(currentDir, potentialDir).replace(/\\/g, '/').replace(/ /g, '%20');
                    return `[${text}](${newPath}/${anchorPart})`;
                }
                
                // Check if the target file exists
                if (!fs.existsSync(resolvedPath)) {
                    // Try to find a similar file by removing special characters from the filename
                    const dirPath = path.dirname(resolvedPath);
                    const fileName = path.basename(resolvedPath);
                    
                    // Remove problematic characters and try to find the file
                    const cleanFileName = fileName.replace(/[\(\)\\]/g, '');
                    const cleanPath = path.join(dirPath, cleanFileName);
                    
                    if (fs.existsSync(cleanPath)) {
                        // Calculate relative path from current file to the cleaned file
                        const newPath = path.relative(currentDir, cleanPath).replace(/\\/g, '/').replace(/ /g, '%20');
                        return `[${text}](${newPath}${anchorPart})`;
                    }
                }
            }
        }
        
        // Also handle local references (same directory)
        if (!decodedLink.includes('/')) {
            const basename = decodedLink.endsWith('.md') ? decodedLink.slice(0, -3) : decodedLink;
            const possibleDir = path.join(currentDir, basename);
            
            if (fs.existsSync(possibleDir) && fs.statSync(possibleDir).isDirectory()) {
                // Re-encode spaces for URL compatibility
                const encodedBasename = basename.replace(/ /g, '%20');
                return `[${text}](${encodedBasename}/${anchorPart})`;
            }
        }
        
        return match;
    }
    
    function walkDir(dir: string): void {
        let files: string[];
        try {
            files = fs.readdirSync(dir);
        } catch (err) {
            console.warn(`Warning: Unable to read directory ${dir}: ${err.message}`);
            return;
        }
        
        for (const file of files) {
            const filePath = path.join(dir, file);
            let stat: fs.Stats;
            
            try {
                stat = fs.statSync(filePath);
            } catch (err) {
                // File might have been moved already, skip it
                continue;
            }
            
            if (stat.isDirectory()) {
                walkDir(filePath);
            } else if (file.endsWith('.md')) {
                let content = fs.readFileSync(filePath, 'utf-8');
                const originalContent = content;
                
                const isIndex = file === 'index.md';
                const currentDir = path.dirname(filePath);
                
                // Update markdown links: [text](path.md)
                const pattern = /\[([^\]]*)\]\(([^)]+\.md)\)/g;
                content = content.replace(pattern, (match, text, link) => {
                    return fixLink(match, text, link, currentDir, isIndex);
                });
                
                if (content !== originalContent) {
                    fs.writeFileSync(filePath, content, 'utf-8');
                    updatesMade.push({
                        message: `Updated references in ${path.relative(docsDir, filePath)}`
                    });
                }
            }
        }
    }
    
    walkDir(docsDir);
    return updatesMade;
}

/**
 * Sync README.md to docs/index.md with necessary path adjustments
 */
function syncReadmeToIndex(projectRoot: string, docsDir: string): FixResult[] {
    const results: FixResult[] = [];
    const readmePath = path.join(projectRoot, 'README.md');
    const indexPath = path.join(docsDir, 'index.md');
    
    if (!fs.existsSync(readmePath)) {
        console.warn('README.md not found in project root');
        return results;
    }
    
    // Read README content
    let content = fs.readFileSync(readmePath, 'utf-8');
    
    // Fix image path (./docs/app.png -> app.png)
    content = content.replace(/src="\.\/docs\/app\.png"/g, 'src="app.png"');
    
    // Fix language links in header
    content = content.replace(/\[English\]\(\.\/README\.md\)/g, '[English](./index.md)');
    content = content.replace(/\.\/docs\/README-ZH_CN\.md/g, './README-ZH_CN.md');
    content = content.replace(/\.\/docs\/README-ZH_TW\.md/g, './README-ZH_TW.md');
    content = content.replace(/\.\/docs\/README\.ru\.md/g, './README.ru.md');
    content = content.replace(/\.\/docs\/README\.ja\.md/g, './README.ja.md');
    content = content.replace(/\.\/docs\/README\.it\.md/g, './README.it.md');
    content = content.replace(/\.\/docs\/README\.es\.md/g, './README.es.md');
    
    // Fix internal documentation links (./docs/User%20Guide -> ./User%20Guide)
    content = content.replace(/\.\/docs\/User%20Guide/g, './User%20Guide');
    content = content.replace(/\.\/docs\/Script%20API/g, './Script%20API');
    content = content.replace(/\.\/docs\/Developer%20Guide/g, './Developer%20Guide');
    
    // Fix specific broken links found in index.md
    // These links point to non-existent files, so we need to fix them
    content = content.replace(/User%20Guide\/quick-start\.md/g, 'User%20Guide/User%20Guide/');
    content = content.replace(/User%20Guide\/installation\.md/g, 'User%20Guide/User%20Guide/Installation%20&%20Setup/');
    content = content.replace(/User%20Guide\/docker\.md/g, 'User%20Guide/User%20Guide/Installation%20&%20Setup/Docker%20installation/');
    content = content.replace(/User%20Guide\/index\.md/g, 'User%20Guide/User%20Guide/');
    content = content.replace(/Script%20API\/index\.md/g, 'Script%20API/');
    content = content.replace(/Developer%20Guide\/index\.md/g, 'Developer%20Guide/Developer%20Guide/');
    content = content.replace(/Developer%20Guide\/contributing\.md/g, 'Developer%20Guide/Developer%20Guide/');
    content = content.replace(/support\/faq\.md/g, 'User%20Guide/User%20Guide/FAQ/');
    content = content.replace(/support\/troubleshooting\.md/g, 'User%20Guide/User%20Guide/');
    
    // Write the adjusted content to docs/index.md
    fs.writeFileSync(indexPath, content, 'utf-8');
    results.push({
        message: `Synced README.md to docs/index.md with path adjustments`
    });
    
    return results;
}

function main(): number {
    // Get the docs directory
    const scriptDir = path.dirname(new URL(import.meta.url).pathname);
    const projectRoot = path.dirname(scriptDir);
    const docsDir = path.join(projectRoot, 'docs');
    
    // Handle Windows paths (remove leading slash if on Windows)
    const normalizedProjectRoot = process.platform === 'win32' && projectRoot.startsWith('/') 
        ? projectRoot.substring(1) 
        : projectRoot;
    const normalizedDocsDir = process.platform === 'win32' && docsDir.startsWith('/') 
        ? docsDir.substring(1) 
        : docsDir;
    
    if (!fs.existsSync(normalizedDocsDir)) {
        console.error(`Error: docs directory not found at ${normalizedDocsDir}`);
        return 1;
    }
    
    console.log(`Fixing MkDocs structure in ${normalizedDocsDir}`);
    console.log('-'.repeat(50));
    
    // Sync README.md to docs/index.md
    const syncResults = syncReadmeToIndex(normalizedProjectRoot, normalizedDocsDir);
    if (syncResults.length > 0) {
        console.log('README sync:');
        for (const result of syncResults) {
            console.log(`  - ${result.message}`);
        }
        console.log();
    }
    
    // Fix duplicate entries
    const fixes = fixDuplicateEntries(normalizedDocsDir);
    if (fixes.length > 0) {
        console.log('Files reorganized:');
        for (const fix of fixes) {
            console.log(`  - ${fix.message}`);
        }
    } else {
        console.log('No duplicate entries found that need fixing');
    }
    
    console.log();
    
    // Update references
    const updates = updateReferences(normalizedDocsDir);
    if (updates.length > 0) {
        console.log('References updated:');
        for (const update of updates) {
            console.log(`  - ${update.message}`);
        }
    } else {
        console.log('No references needed updating');
    }
    
    console.log('-'.repeat(50));
    console.log(`Structure fix complete: ${syncResults.length} README syncs, ${fixes.length} files moved, ${updates.length} files updated`);
    
    return 0;
}

// Run the main function
process.exit(main());